### ISSUE TITLE
Future of BilinearGroupFactory

### ISSUE THREAD

**Issue by JanBobolz** [upbcuk/upb.crypto.math#34]
I want to discuss getting rid of `BilinearGroupFactory`. 

## A history lesson
The idea for the factory came from theory, where we usually just write "generate some type 3 bilinear group for security parameter λ, I don't really care about the details".
Of course, nowadays, we do care about the details: we want to measure performance and be able to say "this scheme takes this long on a BN256 group". So the old vision is actually somewhat outdated.

## The way it actually is right now
From personal experience, using the `BilinearGroupFactory` was almost never really something that made things much easier, it just added one extra step to instantiating the group I really wanted.
For example, to switch from our java bn to mcl, you need to

1. register the mcl provider to the factory
2. guess/check valid parameters for the factory requirements (in particular, security parameter).

## Proposal

What if instead, to switch to mcl, you could just replace
```java
BilinearGroup bg = new BarretoNaehrigProvider(128).provide();
```

with

```java
BilinearGroup bg = new MclBilinearGroupProvider().provide();
```

i.e. we just leave out the `BilinearGroupFactory` middleman and remove this forced specification of requirements.

### Common interface for instantiation
We can still use `BilinearGroupProvider` interface to instantiate bilinear groups. I'd relax the interface slightly to get rid of some of the complexity in using it.

### Security parameter
Instead of requiring the user to supply the desired security level to `provide()`, we could add a `getSecurityLevel()` to `BilinearGroup` to _inform_ users of how many bits security it provides. This makes more sense to me in a world where some instantiations (e.g., mcl's BN256) are not arbitrarily flexible in their security parameter.

### Type
Type 1/2/3 is something users are usually aware of, anyway, when instantiating the group. Also here, the `BilinearGroup` should, in my opinion, inform about the type, instead of making the user specify it. 
(Though in this case, one can argue that it would be nice to prevent a user to accidentally do benchmarking with the wrong group type?! In the worst case, we could add the type as a `provide()` parameter, though it will probably be more confusing than helpful.)

### Hash function requirements
Let's remove these requirements for "I need a hash into G1" and let's just supply them if they exist, otherwise throw an exception when the user tries to access them. 

### Number of prime factors
This should become a constructor argument for those `BilinearGroupProvider`s that actually support composite order groups.

### Discoverability

To replace the "one stop shop" function of the `BilinearGroupFactory` (only one thing you need to know to get either our supersingular type 1 or our type 3 pairing), we could just name our packages better and just encode the "type1/type3" distinction into the package names. Then people will be able to easily find the corresponding classes.
Also, we only have two implementations, so we could just list them exactly where we currently nudge people towards the `BilinearGroupFactory`.

**Comment by feidens** [comment]
Very good and thought-out proposal. As a discussion point I would like to ask if the old target group is still covered. We envisioned that the old target group does not really know what all the curves are, but they know their own requirements (type, hash? , security level) coming from the application level.

**Comment by JanBobolz** [comment]
Less covered than before. I imagine we'd direct them to a table somewhere on our documentation website that lists all available `BilinearGroup`s with their characteristics and let them choose according to their requirements from this list.

To be honest, we were never _really_ able to cater to that target audience, because while functionally, they got what they needed, to actually do something _useful_ with their finished program, they had to manually switch to mcl anyway. For this, they always had to look at some sort of website to find out that mcl exists.

So overall: slightly worse than before, I guess, but not really. We can compensate with good documentation.

**Comment by feidens** [comment]
I see. I conclude for this discussion point that there is no need to generalise the api any more to support the old way of thinking.
Overall, I'm all for the proposal at the top.

**Comment by rheitjoh** [comment]
> Instead of requiring the user to supply the desired security level to provide(), we could add a getSecurityLevel() to BilinearGroup to inform users of how many bits security it provides. This makes more sense to me in a world where some instantiations (e.g., mcl's BN256) are not arbitrarily flexible in their security parameter.

So this would also hold for our groups where we instantiate the group size dynamically depending on the desired security parameter, i.e. which are arbitrarily flexible? Hence, we would change, for example, our BN implementation to work with fixed security parameters, such as offering one instantiation for 80 bits, one for 128 bits, etc. (whatever makes sense to have)? This would also get rid of our [problem with group bit size selection for BN](https://github.com/upbcuk/upb.crypto.math/issues/32).

**Comment by rheitjoh** [comment]
Continuing my previous comment, if we remove the security parameter, I think it makes sense to switch away from the dynamic initialization method completely, and instead use fixed parameters selected by experts instead. I would probably start off with a BN provider for 100 bit security as that allows switching out for Mcl easily (would try to use the same parameters as Mcl uses here).

In general, I can't say I know anything about the limitations of the dynamic parameter selection method, though.

**Comment by rheitjoh** [comment]
Deleted my previous comments to get it all into one, hopefully more cohesive post about the planned implementation of Jan's proposal.

Regarding directory structure, I would propose the following:
```
    pairings/
        type1/
            supersingular/
        type2/
        type3/
            bn/
        debug/
            count/
        generic/
            BilinearGroup
            BilinearGroupImpl
            BilinearGroupProvider
```
There is no point in keeping the `factory` folder since it does not contain a factory anymore, and moving the interfaces to the `pairings` folder makes sense since they are used there. I would put them into the `generic` folder since the other interfaces and abstract classes related to pairings are in there as well.  As desired, there are folders for each type of pairing plus some miscellanous stuff like the debug and counting groups.

As you can see, `BilinearGroupFactory` would be removed, which then leads to the removal of `BilinearGroupRequirement` as well.

`BilinearGroupProvider` would contain the following methods:
- `provide()`: provides the bilinear group as usual (just shorter method name)
- `provideImpl()`: provides the bilinear group impl as usual l (just shorter method name)
- `getSecurityLevel()`: returns the configured security parameter
- `getPairingType()`: returns the configured pairing type
- `isHashIntoG1Supported()`: returns whether the provided pairing would support a hash into G1
- `isHashIntoG2Supported()`: returns whether the provided pairing would support a hash into G2
- `isHashIntoGTSupported()`: returns whether the provided pairing would support a hash into GT

As suggested by Jan, we can configure the number of prime factors for composite group order via the provider's constructor. This has the advantage of having no impact on the interface itself.
Additionally, we can use the constructor to configure stuff like the security parameter and the pairing type for the debug group since that supports all types.
This also allows us to keep the dynamic initialization variants of the BN and Supersingular pairings since we can just give the security parameter to the provider constructor. We can additionally still offer providers with fixed parameters, of course (if that is advantageous to choosing parameters dynamically). If we want to keep offering the dynamic BN variant, the [question of how to deal with group bit size calculation based on the given security parameter](https://github.com/upbcuk/upb.crypto.math/issues/32) still stands.

For example, a prime order debug group might be initialized as follows:
```Java
BilinearGroup bilGroup = new DebugBilinearGroupProvider(100, BilinearGroup.Type.TYPE_1).provide();
```
Or a 128 bit security BN pairing:
```Java
BilinearGroup bilGroup = new BarretoNaehrigProvider(128).provide();
```

**Comment by JanBobolz** [comment]
Just to update: as in #36, there is not necessarily a `BilinearGroupProvider` anymore, the default way is now to instantiate the `BilinearGroup` directly via constructor call.

**Comment by rheitjoh** [comment]
Closed by #36
